src/reader/warming.rs

```diff
@@ -0,0 +1,341 @@
+use std::{
+    collections::HashSet,
+    sync::{Arc, Mutex, Weak},
+    thread::JoinHandle,
+    time::Duration,
+};
+
+use crate::{Executor, Searcher, SearcherIndexGeneration, TantivyError};
+use crate::{Inventory, TrackedObject};
+
+pub const GC_INTERVAL: Duration = Duration::from_secs(1);
+
+/// `Warmer` can be used to maintain segment-level state e.g. caches.
+///
+/// They must be registered with the [IndexReaderBuilder].
+pub trait Warmer: Sync + Send {
+    /// Perform any warming work using the provided [Searcher].
+    fn warm(&self, searcher: &Searcher) -> crate::Result<()>;
+
+    /// Discard internal state for any [SearcherIndexGeneration] not provided.
+    fn garbage_collect(&self, live_generations: &[TrackedObject<SearcherIndexGeneration>]);
+}
+
+/// Warming-related state with interior mutability.
+#[derive(Clone)]
+pub(crate) struct WarmingState(Arc<Mutex<WarmingStateInner>>);
+
+impl WarmingState {
+    pub fn new(
+        num_warming_threads: usize,
+        warmers: Vec<Weak<dyn Warmer>>,
+        searcher_generation_inventory: Inventory<SearcherIndexGeneration>,
+    ) -> crate::Result<Self> {
+        Ok(Self(Arc::new(Mutex::new(WarmingStateInner {
+            num_warming_threads,
+            warmers,
+            gc_thread: None,
+            searcher_generation_ids: Default::default(),
+            searcher_generation_inventory,
+        }))))
+    }
+
+    /// Start tracking a new generation of [Searcher], and [Warmer::warm] it if there are active warmers.
+    ///
+    /// A background GC thread for [Warmer::garbage_collect] calls is uniquely created if there are active warmers.
+    pub fn warm_new_searcher_generation(&self, searcher: &Searcher) -> crate::Result<()> {
+        self.0
+            .lock()
+            .unwrap()
+            .warm_new_searcher_generation(searcher, &self.0)
+    }
+
+    #[cfg(test)]
+    fn gc_maybe(&self) -> bool {
+        self.0.lock().unwrap().gc_maybe()
+    }
+}
+
+struct WarmingStateInner {
+    num_warming_threads: usize,
+    warmers: Vec<Weak<dyn Warmer>>,
+    gc_thread: Option<JoinHandle<()>>,
+    searcher_generation_ids: HashSet<u64>,
```

Comment:
Maybe `warmed_generation_ids` is more descriptive? (and update `warm_new_searcher_generation()` to only insert into it after warmup...)

Reply:
Very good point.

Reply:
I moved it a little bit down...
But not after the warm up :).
Here is my train of thoughts.

This list is here as an optimization to avoid triggers Warmer GC when we identify this is not useful.
In other words:
If we have too many elements in it, this is not a bug.
If we have not enough elements in it, it can induce a bug.

The invariant we want to keep is "we warmed up a generation => that generation id is in that list" (where => means implies).

The invariant require us to add stuff in the list before we warm up.

Being strict on that kind of invariant is useful when doing multithreading or if there is a state that is persisted.
Here I think we don't have this problem, so this is just a defensive practise.

Reply:
I moved it a little bit down...
But not after the warm up :).
Here is my train of thoughts.

This list is here as an optimization to avoid triggers Warmer GC when we identify this is not useful.
In other words:
If we have too many elements in it, this is not a bug.
If we have not enough elements in it, it can induce a bug.

The invariant we want to keep is "we warmed up a generation => that generation id is in that list" (where => means implies).

The invariant require us to add stuff in the list before we warm up.

Being strict on that kind of invariant is useful when doing multithreading or if there is a state that is persisted.
Here I think we don't have this problem, so this is just a defensive practise.